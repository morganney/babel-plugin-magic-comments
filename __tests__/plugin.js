import { resolve, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'

import { jest } from '@jest/globals'
import { transformFileSync } from '@babel/core'

import plugin from '../src/index.js'

const directory = dirname(fileURLToPath(import.meta.url))

jest.mock('magic-comments', () => {
  return {
    ...jest.requireActual('magic-comments'),
    getMagicComment: jest.fn(() => '')
  }
})

describe('babel-plugin-magic-comments', () => {
  it('adds webpackChunkName magic comments', () => {
    const { code: basic } = transformFileSync(
      resolve(directory, '__fixtures__/basic.js'),
      {
        plugins: [
          [
            plugin,
            {
              verbose: true,
              match: 'import',
              webpackChunkName: '**/module.js'
            }
          ]
        ]
      }
    )

    expect(basic).toMatchSnapshot()

    const { code: dynamic } = transformFileSync(
      resolve(directory, '__fixtures__/dynamic.js'),
      {
        plugins: [
          [
            plugin,
            {
              webpackChunkName: true
            }
          ]
        ]
      }
    )

    expect(dynamic).toMatchSnapshot()

    const { code: file } = transformFileSync(resolve(directory, '__fixtures__/file.js'), {
      plugins: [
        [
          plugin,
          {
            webpackChunkName: true
          }
        ]
      ]
    })

    expect(file).toMatchSnapshot()
  })

  it('adds the other magic comments', () => {
    const { code } = transformFileSync(resolve(directory, '__fixtures__/basic.js'), {
      plugins: [
        [
          plugin,
          {
            webpackPrefetch: true,
            webpackPreload: true,
            webpackMode: 'eager',
            webpackFetchPriority: 'high',
            webpackIgnore: true,
            webpackExports: () => ['foo', 'bar'],
            webpackInclude: /\.json$/,
            webpackExclude: (modulePath, importPath) => {
              if (importPath.includes('folder/module')) {
                return /\.skip\.json$/
              }
            }
          }
        ]
      ]
    })

    expect(code).toMatchSnapshot()
  })

  it('does not add magic comments when there are none', () => {
    const { code } = transformFileSync(resolve(directory, '__fixtures__/basic.js'), {
      plugins: [
        [
          plugin,
          {
            webpackChunkName: () => ''
          }
        ]
      ]
    })

    expect(code).toMatchSnapshot()
  })
})
