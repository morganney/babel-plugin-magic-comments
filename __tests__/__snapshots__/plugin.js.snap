// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-plugin-magic-comments adds the other magic comments 1`] = `
""use strict";

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const foo = Promise.resolve().then(() => _interopRequireWildcard(require( /* webpackPrefetch: true, webpackPreload: true, webpackMode: "eager", webpackFetchPriority: "high", webpackIgnore: true, webpackExports: ["foo", "bar"], webpackInclude: /\\.json$/, webpackExclude: /\\.skip\\.json$/ */'./folder/module.js')));"
`;

exports[`babel-plugin-magic-comments adds webpackChunkName magic comments 1`] = `
""use strict";

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const foo = Promise.resolve().then(() => _interopRequireWildcard(require( /* webpackChunkName: "folder-module" */'./folder/module.js')));"
`;

exports[`babel-plugin-magic-comments adds webpackChunkName magic comments 2`] = `
""use strict";

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
(specifier => new Promise(r => r(specifier)).then(s => _interopRequireWildcard(require(s))))( /* webpackChunkName: "lang-[request]" */\`./lang/\${locale}.json\`);
const tr = (specifier => new Promise(r => r(specifier)).then(s => _interopRequireWildcard(require(s))))( /* webpackChunkName: "[request]-to" */\`./\${path}/to.json\`);
(specifier => new Promise(r => r(specifier)).then(s => _interopRequireWildcard(require(s))))( /* webpackChunkName: "one-two-[request]" */\`./one/two/\${three}.json\`);"
`;

exports[`babel-plugin-magic-comments adds webpackChunkName magic comments 3`] = `
""use strict";

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
Promise.resolve().then(() => _interopRequireWildcard(require( /* webpackChunkName: "folder-module" */'./folder/module.js'))).then(() => {
  const slug = 'module';
  const json = \`foo-bar-baz-\${(specifier => new Promise(r => r(specifier)).then(s => _interopRequireWildcard(require(s))))( /* webpackChunkName: "[request]" */\`./\${slug}.json\`)}abc\`;
  return json;
});
Promise.resolve().then(() => _interopRequireWildcard(require( /* some comment */'./folder/skip.js')));
reg([{
  module: Promise.resolve().then(() => _interopRequireWildcard(require( /* webpackChunkName: "@pkg-foo" */'@pkg/foo'))),
  elem: 'Foo'
}, {
  module: Promise.resolve().then(() => _interopRequireWildcard(require( /* webpackChunkName: "@pkg-bar" */'@pkg/bar'))),
  elem: 'Bar'
}, {
  module: Promise.resolve().then(() => _interopRequireWildcard(require( /* webpackChunkName: "@pkg-baz" */'@pkg/baz'))),
  elem: 'Baz'
}]);
Promise.all([Promise.resolve().then(() => _interopRequireWildcard(require( /* webpackChunkName: "@pkg-a" */'@pkg/a'))), Promise.resolve().then(() => _interopRequireWildcard(require( /* webpackChunkName: "@pkg-b" */'@pkg/b'))), Promise.resolve().then(() => _interopRequireWildcard(require( /* webpackChunkName: "@pkg-c" */'@pkg/c')))]);"
`;

exports[`babel-plugin-magic-comments does not add magic comments when there are none 1`] = `
""use strict";

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
const foo = Promise.resolve().then(() => _interopRequireWildcard(require('./folder/module.js')));"
`;
